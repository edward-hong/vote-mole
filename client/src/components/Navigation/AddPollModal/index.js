import React from 'react'
import { connect } from 'react-redux'
import { reduxForm, Field, FieldArray } from 'redux-form'
import { withRouter } from 'react-router-dom'
import {
	Modal,
	ModalHeader,
	ModalBody,
	ModalFooter,
	Button,
	Form,
} from 'reactstrap'
import { ajax } from 'rxjs/ajax'
import { compose, or, map, equals } from 'ramda'

import AddPollQuestionInput from './AddPollQuestionInput'
import AddPollOptionsInputs from './AddPollOptionsInputs'
import BlackButton from '../../styled/BlackButton'

const AddPollModal = ({
	isOpen,
	toggle,
	reset,
	handleSubmit,
	pristine,
	submitting,
	auth,
	history,
}) => {
	const cancel = () => {
		reset()
		toggle()
	}

	// Submit poll to backend
	const submit = ({ pollQuestion, options }) =>
		ajax({
			url: '/poll/submit',
			method: 'POST',
			headers: {
				'Content-Type': 'application/json',
			},
			body: {
				userID: auth._id,
				pollQuestion: pollQuestion,
				pollOptions: map(option => ({ option }), options),
			},
		}).subscribe(
			({ response }) => {
				cancel()
				// history.push(`/poll/${response}`)
			},
			({ status }) => {
				if (equals(status, 406)) {
					toggle()
				}
			}
		)

	return (
		<Modal isOpen={isOpen} toggle={toggle}>
			<ModalHeader toggle={cancel}>Add Poll</ModalHeader>
			<ModalBody>
				<Form onSubmit={handleSubmit(submit)} id="add_poll">
					<Field
						name="pollQuestion"
						type="text"
						label="Poll Question"
						placeholder="Enter your poll question"
						component={AddPollQuestionInput}
					/>
					<FieldArray name="options" component={AddPollOptionsInputs} />
				</Form>
			</ModalBody>
			<ModalFooter>
				<Button outline type="submit" form="add_poll" color="success">
					Submit
				</Button>{' '}
				<Button
					outline
					color="secondary"
					disabled={or(pristine, submitting)}
					onClick={reset}>
					Clear Values
				</Button>
				<BlackButton outline onClick={cancel}>
					Cancel
				</BlackButton>
			</ModalFooter>
		</Modal>
	)
}

// Validate add poll form
const validate = ({ pollQuestion, options }) => {
	const errors = {}

	if (!pollQuestion) errors.pollQuestion = 'Poll question is required'

	if (!options || options.length < 2) {
		errors.options = ['Poll must have at least 2 options']
	} else if (options && options.length > 20) {
		errors.options = ['No more than 20 options allowed']
	} else {
		const optionsArrayErrors = []
		options.forEach((option, i) => {
			if (!option || !option.length) {
				optionsArrayErrors[i] = 'Please fill out all options fields'
			}
		})
		if (optionsArrayErrors.length) {
			errors.options = optionsArrayErrors
		}
	}

	if (options && options[0]) {
		const unique = [...new Set(options)]
		if (unique.length !== options.length) {
			errors.options = ['Options must be unique']
		}
	}

	return errors
}

const mapStateToProps = ({ form, auth }) => ({ form, auth })

export default compose(
	reduxForm({ form: 'add_poll', validate }),
	connect(mapStateToProps),
	withRouter
)(AddPollModal)
