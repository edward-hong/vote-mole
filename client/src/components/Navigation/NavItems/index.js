import React from 'react'
import { connect } from 'react-redux'
import { NavLink, NavItem } from 'reactstrap'
import styled from 'styled-components'
import { isNil } from 'ramda'

import { AUTH_LOGOUT_PATH } from '../../../constants'

const NavLinkItem = styled(NavItem)`
	text-align: center;
`

// Render nothing if user info hasn't been fetched from backend
// Then conditionally render NavItems depending on whether user is logged in
const NavItems = ({ toggleLogin, toggleAddPoll, isLoggedIn }) =>
	isNil(isLoggedIn) ? null : isLoggedIn ? (
		<>
			<NavLink href="#" onClick={toggleAddPoll}>
				<NavLinkItem>Add Poll</NavLinkItem>
			</NavLink>
			<NavLink href={AUTH_LOGOUT_PATH}>
				<NavLinkItem>Logout</NavLinkItem>
			</NavLink>
		</>
	) : (
		<NavLink href="#" onClick={toggleLogin}>
			<NavLinkItem>Login</NavLinkItem>
		</NavLink>
	)

const mapStateToProps = ({ auth }) => ({ isLoggedIn: auth })

export default connect(mapStateToProps)(NavItems)
