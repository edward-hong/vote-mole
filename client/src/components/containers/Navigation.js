import React, { useState } from 'react'
import { Link } from 'react-router-dom'
import { not } from 'ramda'
import {
	Collapse,
	Navbar,
	NavbarToggler,
	NavbarBrand,
	Nav,
	NavItem,
	NavLink,
} from 'reactstrap'
import styled from 'styled-components'

import { HOME_PATH, COLOURS } from '../../constants'

const NavigationBar = styled(Navbar)`
	margin-bottom: 2rem;
	border-bottom: 1px solid ${COLOURS.primary};
`

const Hamburger = styled(NavbarToggler)`
	&& {
		color: ${COLOURS.primary};
		border-color: ${COLOURS.primary};
		width: 56px;
		height: 40px;
		position: relative;
		line-height: 50px;

		span {
			background-image: none;
			width: 30px;
			height: 2px;
			background-color: ${({ active }) =>
				active ? 'transparent' : COLOURS.primary};
			position: absolute;
			transform: translate(-50%, -50%);

			&:before,
			&:after {
				content: '';
				position: absolute;
				width: 30px;
				height: 2px;
				background-color: ${COLOURS.primary};
				left: -0.0001%;
				transition: 0.2s;
			}

			&:before {
				top: ${({ active }) => (active ? 0 : '-8px')};
				transform: ${({ active }) => (active ? 'rotate(135deg)' : 'none')};
			}

			&:after {
				top: ${({ active }) => (active ? 0 : '8px')};
				transform: ${({ active }) => (active ? 'rotate(45deg)' : 'none')};
			}
		}
	}
`

const NavLinkItem = styled(NavItem)`
	text-align: center;
`

const Navigation = () => {
	const [expandNav, setExpandNav] = useState(false)

	const handleToggleNav = () => setExpandNav(not(expandNav))

	return (
		<NavigationBar light expand="md">
			<NavbarBrand tag={Link} to={HOME_PATH}>
				<img
					src="https://res.cloudinary.com/avatarhzh/image/upload/v1509887327/build-a-voting-app/logo.svg"
					alt="vote mole logo"
				/>
				VoteMole
			</NavbarBrand>
			<Hamburger active={expandNav} onClick={handleToggleNav} />
			<Collapse isOpen={expandNav} navbar>
				<Nav navbar>
					<NavLink href="#">
						<NavLinkItem>Login</NavLinkItem>
					</NavLink>
				</Nav>
			</Collapse>
		</NavigationBar>
	)
}

export default Navigation
