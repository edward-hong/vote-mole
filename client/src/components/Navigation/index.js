import React, { useState } from 'react'
import { Link } from 'react-router-dom'
import { not } from 'ramda'
import { Collapse, Navbar, NavbarToggler, NavbarBrand, Nav } from 'reactstrap'
import styled from 'styled-components'

import NavItems from './NavItems'
import LoginModal from './LoginModal'
import AddPollModal from './AddPollModal'
import { HOME_PATH, COLOURS } from '../../constants'

const NavigationBar = styled(Navbar)`
	margin-bottom: 2rem;
	border-bottom: 1px solid ${COLOURS.primary};
`

const CollapsibleItems = styled(Collapse)`
	flex-direction: row-reverse;
`

const Hamburger = styled(NavbarToggler)`
	&& {
		color: ${COLOURS.primary};
		border-color: ${COLOURS.primary};
		width: 56px;
		height: 40px;
		position: relative;
		line-height: 50px;

		/* hamburger animation */
		span {
			background-image: none;
			width: 30px;
			height: 2px;
			background-color: ${({ active }) =>
				active ? 'transparent' : COLOURS.primary};
			position: absolute;
			transform: translate(-50%, -50%);

			&:before,
			&:after {
				content: '';
				position: absolute;
				width: 30px;
				height: 2px;
				background-color: ${COLOURS.primary};
				left: -0.0001%;
				transition: 0.2s;
			}

			&:before {
				top: ${({ active }) => (active ? 0 : '-8px')};
				transform: ${({ active }) => (active ? 'rotate(135deg)' : 'none')};
			}

			&:after {
				top: ${({ active }) => (active ? 0 : '8px')};
				transform: ${({ active }) => (active ? 'rotate(45deg)' : 'none')};
			}
		}
	}
`

const Navigation = () => {
	const [expandNav, setExpandNav] = useState(false)
	const [isOpenLogin, setIsOpenLogin] = useState(false)
	const [isOpenAddPoll, setIsOpenAddPoll] = useState(false)

	const handleToggle = (state, setState) => () => setState(not(state))

	return (
		<NavigationBar light expand="md">
			{/* Logo and App Name */}
			<NavbarBrand tag={Link} to={HOME_PATH}>
				<img
					src="https://res.cloudinary.com/avatarhzh/image/upload/v1509887327/build-a-voting-app/logo.svg"
					alt="vote mole logo"
				/>
				VoteMole
			</NavbarBrand>

			{/* Hamburger icon for smaller screens */}
			<Hamburger
				active={expandNav ? 1 : 0}
				onClick={handleToggle(expandNav, setExpandNav)}
			/>
			<CollapsibleItems isOpen={expandNav} navbar>
				<Nav navbar>
					<NavItems
						toggleLogin={handleToggle(isOpenLogin, setIsOpenLogin)}
						toggleAddPoll={handleToggle(isOpenAddPoll, setIsOpenAddPoll)}
					/>
				</Nav>
			</CollapsibleItems>
			<LoginModal
				isOpen={isOpenLogin}
				toggle={handleToggle(isOpenLogin, setIsOpenLogin)}
			/>
			<AddPollModal
				isOpen={isOpenAddPoll}
				toggle={handleToggle(isOpenAddPoll, setIsOpenAddPoll)}
			/>
		</NavigationBar>
	)
}

export default Navigation
