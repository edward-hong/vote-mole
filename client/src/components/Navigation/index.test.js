import React from 'react'
import { fireEvent } from '@testing-library/react'

import Navigation from '.'
import renderWithRouterAndRedux from '../../utils/test/renderWithRouterAndRedux'

describe('Navigation', () => {
	it('renders correctly', () => {
		const { container } = renderWithRouterAndRedux(<Navigation />)

		expect(container.firstChild).toMatchSnapshot()
	})

	it('renders correctly when nav is expanded', () => {
		const { container } = renderWithRouterAndRedux(<Navigation />)
		const hamburger = container.getElementsByTagName('button')[0]

		fireEvent.click(hamburger)

		expect(container.firstChild).toMatchSnapshot()
	})
})
