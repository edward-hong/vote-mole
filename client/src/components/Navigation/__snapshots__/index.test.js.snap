// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navigation renders correctly 1`] = `
.c0 {
  margin-bottom: 2rem;
  border-bottom: 1px solid #21252a;
}

.c2 {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c1.c1 {
  color: #21252a;
  border-color: #21252a;
  width: 56px;
  height: 40px;
  position: relative;
  line-height: 50px;
}

.c1.c1 span {
  background-image: none;
  width: 30px;
  height: 2px;
  background-color: #21252a;
  position: absolute;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c1.c1 span:before,
.c1.c1 span:after {
  content: '';
  position: absolute;
  width: 30px;
  height: 2px;
  background-color: #21252a;
  left: -0.0001%;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c1.c1 span:before {
  top: -8px;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

.c1.c1 span:after {
  top: 8px;
  -webkit-transform: none;
  -ms-transform: none;
  transform: none;
}

<nav
  class="c0 navbar navbar-expand-md navbar-light"
>
  <a
    class="navbar-brand"
    href="/"
  >
    <img
      alt="vote mole logo"
      src="https://res.cloudinary.com/avatarhzh/image/upload/v1509887327/build-a-voting-app/logo.svg"
    />
    VoteMole
  </a>
  <button
    active="0"
    class="c1 navbar-toggler"
    type="button"
  >
    <span
      class="navbar-toggler-icon"
    />
  </button>
  <div
    class="c2 collapse navbar-collapse"
  >
    <ul
      class="navbar-nav"
    />
  </div>
</nav>
`;

exports[`Navigation renders correctly when nav is expanded 1`] = `
.c0 {
  margin-bottom: 2rem;
  border-bottom: 1px solid #21252a;
}

.c2 {
  -webkit-flex-direction: row-reverse;
  -ms-flex-direction: row-reverse;
  flex-direction: row-reverse;
}

.c1.c1 {
  color: #21252a;
  border-color: #21252a;
  width: 56px;
  height: 40px;
  position: relative;
  line-height: 50px;
}

.c1.c1 span {
  background-image: none;
  width: 30px;
  height: 2px;
  background-color: transparent;
  position: absolute;
  -webkit-transform: translate(-50%,-50%);
  -ms-transform: translate(-50%,-50%);
  transform: translate(-50%,-50%);
}

.c1.c1 span:before,
.c1.c1 span:after {
  content: '';
  position: absolute;
  width: 30px;
  height: 2px;
  background-color: #21252a;
  left: -0.0001%;
  -webkit-transition: 0.2s;
  transition: 0.2s;
}

.c1.c1 span:before {
  top: 0;
  -webkit-transform: rotate(135deg);
  -ms-transform: rotate(135deg);
  transform: rotate(135deg);
}

.c1.c1 span:after {
  top: 0;
  -webkit-transform: rotate(45deg);
  -ms-transform: rotate(45deg);
  transform: rotate(45deg);
}

<nav
  class="c0 navbar navbar-expand-md navbar-light"
>
  <a
    class="navbar-brand"
    href="/"
  >
    <img
      alt="vote mole logo"
      src="https://res.cloudinary.com/avatarhzh/image/upload/v1509887327/build-a-voting-app/logo.svg"
    />
    VoteMole
  </a>
  <button
    active="1"
    class="c1 navbar-toggler"
    type="button"
  >
    <span
      class="navbar-toggler-icon"
    />
  </button>
  <div
    class="c2 collapsing navbar-collapse"
    style="height: 0px;"
  >
    <ul
      class="navbar-nav"
    />
  </div>
</nav>
`;
